### PR TITLE
Implements IPv6 support

### PR DESCRIPTION
**Pull request by frenzis01**
Solves issue #1.
Refactored code in `NwInterface.cpp` to avoid code duplication and to merge ipv4 and ipv6 "shareable" sections.

### CODE REVIEW

**Comment by simonepassera** [comment]
What is the input file for stats script ?

**Review by simonepassera — APPROVED** [review]

**Comment by yuricaprini** [comment]
Code seems to detect IPv6 packets , but doesn't apply policy to IPv4 properly. You can test it manually by yourself:
- run ipt_geofence as always (but with -v option enabled) on one terminal tab
- run curl "pravo.gov.ru" on a second terminal tab

expected behaviour: ipt_geofence should apply DROP policy to that connection printing feedback on stdout and curl should not finish its execution. (This is the behaviour obtained using the original ntop:main repo)

actual behaviour: curl command resolves. ipt_geofence does not give any feedback about that connection (PASS neither)

It's just a simple and fast manual test, but needs some further investigation and maybe we should start to consider building a real test routine.. 

Let me know what you think about it

**Comment by yuricaprini** [comment]
> Code seems to detect IPv6 packets , but doesn't apply policy to IPv4 properly. You can test it manually by yourself:
> 
> * run ipt_geofence as always (but with -v option enabled) on one terminal tab
> * run curl "pravo.gov.ru" on a second terminal tab
> 
> expected behaviour: ipt_geofence should apply DROP policy to that connection printing feedback on stdout and curl should not finish its execution. (This is the behaviour obtained using the original ntop:main repo)
> 
> actual behaviour: curl command resolves. ipt_geofence does not give any feedback about that connection (PASS neither)
> 
> It's just a simple and fast manual test, but needs some further investigation and maybe we should start to consider building a real test routine..
> 
> Let me know what you think about it

I have noticed that in sample_config.json you have changed default policy and also added a new, different, continent branch (shouldn't there be just one?). However, even restoring the previous configuration,  the problem mentioned continues to persist.

File: ipt_config_utils/single_iface.sh [hunk 1]
```diff
@@ -21,40 +22,49 @@ conntrack -U --mark 0 > /dev/null 2>&1
 
 # Routing needed by DNAT (DNS)
 echo 1 > /proc/sys/net/ipv4/ip_forward
+echo 1 > /proc/sys/net/ipv6/conf/all/forwarding 
+
+$IPTABLES -F
+$IPTABLES -t nat -F
+$IPTABLES -t mangle -F
```
[the remark this answers is not in the document]
Comment (frenzis01) [reply ipt_config_utils/single_iface.sh:30]: You are right, fixed.

**Comment by frenzis01** [comment]
> Code seems to detect IPv6 packets , but doesn't apply policy to IPv4 properly. You can test it manually by yourself:
> 
>     * run ipt_geofence as always (but with -v option enabled) on one terminal tab
> 
>     * run curl "pravo.gov.ru" on a second terminal tab
> 
> 
> expected behaviour: ipt_geofence should apply DROP policy to that connection printing feedback on stdout and curl should not finish its execution. (This is the behaviour obtained using the original ntop:main repo)
> 
> actual behaviour: curl command resolves. ipt_geofence does not give any feedback about that connection (PASS neither)
> 
> It's just a simple and fast manual test, but needs some further investigation and maybe we should start to consider building a real test routine..
> 
> Let me know what you think about it

There were some problems when dissecting packets that screwed up the src/dst ports displayed. There was also a duplicated byte ordering conversion when checking if an address was private or not.

**Comment by lucaderi** [comment]
@frenzis01 Do you think it's time to merge this PR ?

**Comment by frenzis01** [comment]
> @frenzis01 Do you think it's time to merge this PR ?

Yes. Everything should be fine now

File: NwInterface.cpp [hunk 2]
```diff
@@ -155,56 +155,63 @@ Marker NwInterface::dissectPacket(const u_char *payload, u_int payload_len) {
...
+    char src[INET6_ADDRSTRLEN] = {}, dst[INET6_ADDRSTRLEN] = {};
+
+    if (iph->version == 6) {
+      ipv6 = true;
+      struct ip6_hdr *ip6h = (struct ip6_hdr *)&payload[ip_offset];
+      proto = ip6h->ip6_nxt;
+
+      // ipv6 address stringification
+      inet_ntop(AF_INET6, &(ip6h->ip6_src), src, sizeof(src));
+      inet_ntop(AF_INET6, &(ip6h->ip6_src), dst, sizeof(dst));
```
Comment (yuricaprini) [inline NwInterface.cpp:175]:
```suggestion
      inet_ntop(AF_INET6, &(ip6h->ip6_dst), dst, sizeof(dst));
```

**Review by yuricaprini — APPROVED** [review]
Everything looks good to me. Time to merge! 😃 🎉

**Comment by lucaderi** [comment]
Thank you all

**Pull request merged**